src/interface.jl

```diff
@@ -216,56 +227,75 @@ has_wings(duck::Duck)::Bool
 """
 macro implement(can_type, by, sig)
 
-    func_name, func_arg_names, func_arg_types, return_type =
+    func_name, func_arg_names, func_arg_types, kwarg_names, return_type =
         parse_implement(can_type, by, sig)
     # @info "sig" func_name func_arg_names func_arg_types
 
     # generate code
     expr = quote
         function $func_name end
         BinaryTraits.register($can_type, $func_name,
-            ($(func_arg_types...),), $return_type)
+                ($(func_arg_types...),), tuple(($kwarg_names)...), $return_type)
     end
     display_expanded_code(expr)
     return esc(expr)
 end
 
 # Parsing function for @implement macro
 function parse_implement(can_type, by, sig)
-    usage = "Invalid @implement usage."
-    if !(can_type isa Symbol) || by !== :by || !(sig isa Expr)
-        throw(SyntaxError(usage))
-    end
+    usage = "usage: @implement <Type> by <function specification::<ReturnType>"
+    can_type isa Symbol && sig isa Expr && by === :by || throw(SyntaxError(usage))
 
     # Is return type specified?
-    has_return_type = sig.head == Symbol("::")
-    if has_return_type
+    if sig.head === Symbol("::") 
         return_type = sig.args[2]
         sig = sig.args[1]
     else
         return_type = :Any
     end
 
-    # TODO should we use @assert here?
-    @assert sig isa Expr
-    @assert sig.head === :call
+    sig isa Expr && sig.head === :call || throw(SyntaxError(usage))
 
     # parse signature
     func_name = sig.args[1]   # must be Symbol
     func_arg_names = Symbol[]
-    func_arg_types = Symbol[]
-    for (idx, x) in enumerate(sig.args[2:end])  # x must be Expr of 1 or 2 symbols
-        @assert x isa Expr
-        @assert x.head == Symbol("::")
-        @assert x.args |> length in [1,2]
-        if length(x.args) == 1
-            push!(func_arg_names, Symbol("x$idx"))
-            push!(func_arg_types, x.args[1])
-        else
-            push!(func_arg_names, x.args[1])
-            push!(func_arg_types, x.args[2])
+    func_arg_types = Any[]
+    func_kwarg_names = Symbol[]
+    firstarg = 2
+    if length(sig.args) >= 2 && sig.args[2] isa Expr && sig.args[2].head == :parameters
+        for x in sig.args[2].args
+            push!(func_kwarg_names, extract_name(x, nothing))
```

Comment:
Based upon the comment about extract-functions below, this would have to use `QuoteNode`:
```julia
push!(func_kwarg_names, QuoteNode(extract_name(x, nothing)))
```

Reply:
Why would we need the `QuoteNode` in this case? The PR implementation is fine with an array of `Symbol`s.
BTW, the `extract_name` is called recursively in order to handle keyword argument expressions like `kw::Type` and also `kw::Type=defl` and `kw=defl`.

Reply:
Oops, I forgot to mention that I want to get rid of the following `tuple` call when the `@implement` macro is expanded:

```julia
julia> @macroexpand(@implement CanDive by dive(angle::Float64; speed, river = :amazon))
quote
    function dive end
    BinaryTraits.register(CanDive, dive, (Float64,), tuple([:speed, :river]...), Any)
end
```

If you maintain the array of `QuoteNode`'s, then you can replace the `kwarg_names` part of this line:
```julia
BinaryTraits.register($can_type, $func_name,
                ($(func_arg_types...),), tuple(($kwarg_names)...), $return_type)
```
with this:
```julia
BinaryTraits.register($can_type, $func_name,
                ($(func_arg_types...),), ($(kwarg_names...),), $return_type)
```

and it would expand more nicely into:
```julia
quote
    function dive end
    BinaryTraits.register(CanDive, dive, (Float64,), (:speed, :river,), Any)
end
```

Reply:
done

Reply:
I chose a slightly different approach without `QuoteNote`, giving the same macro expansion.